pkg/metrics/builder/builder.go

```diff
@@ -0,0 +1,123 @@
+package builder
+
+import (
+	"errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// MetricBuilder is the basic metric type that must be used by any potential metric added to
+// the system. The build function will use the properties of this struct and create a collector
+// which will be used to push.
+type MetricBuilder struct {
+	// labels contains the list of all label keys that the collector object will have.
+	labels []string
+	// labelKeyValues stores all the label keys and values collected from the installer and
+	// will be used to create the collector object with the stored values.
+	LabelKeyValues map[string]string
+	// desc describes the metric that is being created. This field will be shown on Prometheus
+	// dashboard as a help text.
+	desc string
+	// name describes the name of the metric that needs to be pushed.
+	name string
+	// value stores the value of the metric that needs to be pushed. Collected from the installer.
+	Value float64
+	// buckets keeps a list of bucket values that will be used during the creation of a Histogram
+	// object.
+	buckets []float64
+	// metricType defines what type of a collector object should the PromCollector function return.
+	metricType MetricType
+}
+
+// MetricType defines what types of metrics are
+type MetricType string
+
+const (
+	// Histogram denotes that the type of the collector object should be a Prometheus Histogram.
+	Histogram = "Histogram"
+
+	// Counter denotes that the type of the collector object should be a Prometheus Counter.
+	Counter = "Counter"
+)
+
+// PromCollector function creates the required prometheus collector object with the values
+// it has in the MetricBuilder calling object.
+func (m MetricBuilder) PromCollector() (prometheus.Collector, error) {
+	switch m.metricType {
+	case Counter:
+		return buildCounter(m), nil
+	case Histogram:
+		return buildHistogram(m), nil
+	default:
+		return nil, errors.New("invalid metric builder type. cannot create collector")
+	}
+}
+
+// buildCounter returns a prometheus counter object with the value and labels set
+// in the MetricBuilder object.
+func buildCounter(m MetricBuilder) prometheus.Collector {
```

Comment:
copying the struct because of non-pointer type seems not great, maybe just make it pointer reference...